wukong/request.py

```diff
@@ -27,105 +31,135 @@ class SolrRequest(object):
+        if (self.zookeeper and
+            not is_retry and
+            self._last_request and
+            ((time.time() - self._last_request) / 60) > self.REFRESH_FREQUENCY):
+            self.attempt_zookeeper_refresh()
+
+        request_params = {
+            'wt': 'json',
+            'omitHeader': True,
+            'json.nl': 'map'
+        }
+        if params:
+            request_params.update(params)
+
+        response = None
+        for host in random.sample(self.master_hosts, len(self.master_hosts)):
+            full_path = urljoin(host, path)
             try:
+                logger.debug(
+                    'Sending request to solr. host="%s" path="%s"',
+                    host,
+                    path
+                )
+
+                self._last_request = time.time()
                 response = self.client.request(
                     method,
-                    fullpath,
-                    params=resultparams,
-                    headers=headers,
+                    full_path,
+                    params=request_params,
+                    headers=request_headers,
                     data=body,
                     timeout=self.timeout
                 )
 
-                # Connected to the node, but didn't get a successful response
-                if (len(self.servers) > 0 and
-                        hasattr(response, 'status_code') and
-                        response.status_code != 200):
-                    # try with another node
-                    handle_error()
-
-                return response
-
-            # Didn't successfully connect to the node
-            except ConnectionError as e:
-                if len(self.servers) > 0:
-                    # try with another node
-                    handle_error()
-
-                raise SolrError(str(e))
+                logger.debug(
+                    'Retrieved response from SOLR. host="%s" path="%s" '
+                    'status_code="%s"',
+                    host,
+                    path,
+                    response.status_code
+                )
 
-        self.servers = list(self.current_hosts)
-        if len(self.servers) == 0:
-            handle_error()
+                if response.status_code == 200:
+                    # We've had a successful request. No need to keep trying
+                    break
+                else:
+                    logger.info(
+                        'Unsucessful request to SOLR'
+                        'status_code="%s" reason="%s"',
+                        response.status_code,
+                        response.reason
+                    )
+                    response = None
+
+            except ConnectionError:
+                response = None
+                logger.info(
+                    'Failed to connect to SOLR',
+                    exc_info=True
+                )
 
-        random.shuffle(self.servers)
-        host = self.servers.pop(0)
+        if not response:
```

Comment:
If we've already tried all of the available solr hosts, are we expecting a retry of all of them to be any more successful?

Reply:
The goal here is that if we tried them all, we want to refresh the list in case zookeeper has a new list for us.